Test inter process random state transfers

Dr15Jones opened the pull request:
#### PR description:

Adds a test to show how to use a EDProducer in cmsRun to drive a different EDProducer in a stand alone process where random engine state is passed back and forth.

#### PR validation:

The code compiles and framework unit tests pass.

Dr15Jones committed:
Created RandomNumberGeneratorState

This is used to allow inter-process communication of the random number state.

Dr15Jones committed:
Added ExternalRandomNumberGeneratorService

This provides a RandomNumberGenerator object that can easily be used with the TestProcessor to drive the random numbers in a test or for a non-cmsRun process.

Dr15Jones committed:
Add test for passing random engine state between processes

Created a test that uses SharedMemory package to have a module within cmsRun pass random state to and from another EDProducer run in an external process.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29384/14495

- This PR adds an extra 44KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

DataFormats/Common
DataFormats/TestObjects
FWCore/Integration
FWCore/Services


@makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @rovere, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on DataFormats/Common/interface/RandomNumberGeneratorState.h:
```diff
@@ -0,0 +1,26 @@
+    RandomNumberGeneratorState& operator=(RandomNumberGeneratorState&&) = default;
+
+    std::vector<unsigned long> state_;
+    long seed_;
```
Would encapsulation of the member data make sense here?

Dr15Jones replied:
This isn't a 'real' data format since it never is put in a Principal. It is really an internal detail of how the processes communicate. I put it here since I needed a ROOT dictionary to be generated and this class will be used by the FWCore test as well as the upcoming generator code. That is why I just used a struct.

makortel commented on FWCore/Integration/bin/interprocess_random.cc:
```diff
@@ -0,0 +1,207 @@
+#include <thread>
+
+//TEMP
+//#include "CLHEP/Random/JamesRandom.h"
```
Is this commented code still useful?

Dr15Jones replied:
Thanks

makortel commented on FWCore/Integration/bin/interprocess_random.cc:
```diff
@@ -0,0 +1,207 @@
+    std::string const uniqueID(vm[kUniqueIDOpt].as<std::string>());
+    {
+      //using namespace boost::interprocess;
+      //auto controlNameUnique = unique_name(memoryName, uniqueID);
```
Is this commented code still useful?

makortel commented on FWCore/Integration/bin/interprocess_random.cc:
```diff
@@ -0,0 +1,207 @@
+          edm::ServiceRegistry::createContaining(std::unique_ptr<edm::RandomNumberGenerator>(randomService));
+
+      //CLHEP::HepJamesRandom rng(12345);
+      //randomService->setState(rng.put(), rng.getSeed());
```
Is this commented code still useful?

makortel commented on FWCore/Integration/test/RandomIntProducer.cc:
```diff
@@ -0,0 +1,46 @@
+
+  void RandomIntProducer::produce(edm::Event& iEvent, edm::EventSetup const&) {
+    edm::Service<edm::RandomNumberGenerator> gen;
+    std::cout << gen->getEngine(iEvent.streamID()).name() << std::endl;
```
Can MessageLogger be used here?

Dr15Jones replied:
That can go away altogether. I had just used it as a test to see if the name was useful.

makortel commented on FWCore/Integration/test/TestInterProcessRandomProd.cc:
```diff
@@ -0,0 +1,250 @@
+                                  edm::LuminosityBlockIndex iLumi) {
+      edm::Service<edm::RandomNumberGenerator> gen;
+      //NOTE: root serialize requires a `void*` not a `void const*` even though it doesn't modify the object
+      randSerializer_.serialize(const_cast<edm::RandomNumberGeneratorState&>(iState));
```
Can we change `ROOTSerializer` to take `void const*`? (in a later PR)

Dr15Jones replied:
We can. The underlying problem is ROOT's code which uses the 'void*' instead of the 'void const*'.

makortel replied:
> We can. The underlying problem is ROOT's code which uses the 'void*' instead of the 'void const*'.

I guessed that would be the case. At least the `const_cast` would then be isolated and hidden from the rest...

makortel commented on FWCore/Services/src/ExternalRandomNumberGeneratorService.cc:
```diff
@@ -0,0 +1,89 @@
+
+using namespace edm;
+
+const std::vector<RandomEngineState> s_dummyStates;
```
How about
```suggestion
namespace {
  const std::vector<RandomEngineState> s_dummyStates;
}
```

Dr15Jones committed:
Handle RanecuEngine state setting special

This follows how the RandomNumberGeneratorService treats RanecuEngine when doing replay.

Dr15Jones committed:
Removed unneeded comments and debug printing

Dr15Jones committed:
Put global into anonymous namespace.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29384/14497

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #29384 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+1
Tested at: 064694450cace5874ef86c99e2e30a8617448aff
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-016e2c/5511/summary.html
CMSSW: CMSSW_11_1_X_2020-04-02-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-016e2c/5511/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691786
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5209e297beaa4fbd41f8d11655fbeb5057e7f62c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-016e2c/5515/summary.html
CMSSW: CMSSW_11_1_X_2020-04-02-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-016e2c/5515/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691788
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
